### PR TITLE
Fix SSH/scp/SFTP issues for OpenSSH 9.0+ and Flatcar stable

### PR DESCRIPTION
**Pull request by kopiczko**
This PR tries to fix various issues with Flatcar builds. This fixes multiple issues but at this point it isn't possible to test them in isolation.

Fixes https://github.com/kubernetes-sigs/image-builder/issues/905

Workaround for https://github.com/kubernetes-sigs/image-builder/issues/859

**Additional context**
Add any other context for the reviewers

### CODE REVIEW

File: images/capi/packer/files/bootstrap-flatcar.sh [hunk 1]
```diff
@@ -1,11 +1,16 @@
-#!/bin/bash
-set -e
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
```
Comment (invidian) [inline images/capi/packer/files/bootstrap-flatcar.sh:5]: This changes seem unrelated. Is there a change we could have them committed separately with some reasoning (like aligning to other scripts, aligning to some standard, best practices reference etc.) ?

I'd really appreciate commit per solved issue, so then it's easier to validate that issue X has actually been resolved by commit.

File: images/capi/packer/qemu/packer.json [hunk 2]
```diff
@@ -52,15 +52,18 @@
...
       "extra_arguments": [
+        "--scp-extra-args",
+        "{{user `ansible_scp_extra_args`}}",
         "--extra-vars",
         "{{user `ansible_common_vars`}}",
         "--extra-vars",
         "{{user `ansible_extra_vars`}}"
       ],
       "playbook_file": "./ansible/node.yml",
+      "sftp_command": "{{user `sftp_command`}}",
```
Comment (invidian) [inline images/capi/packer/qemu/packer.json:66]: Is `sftp_command` still needed with `scp` `-O` extra argument?

File: images/capi/packer/qemu/packer.json [hunk 3]
```diff
@@ -137,6 +141,7 @@
     "output_directory": "./output/{{user `build_name`}}-kube-{{user `kubernetes_semver`}}",
     "python_path": "",
     "qemu_binary": "qemu-system-x86_64",
+    "sftp_command": "/usr/lib/sftp-server -e",
```
Comment (invidian) [inline images/capi/packer/qemu/packer.json:144]: Does this refer to local sftp binary or sftp binary on the target VM?

**Review by invidian — COMMENTED** [review]
I'm testing building and I get this error right now:
`    sig-flatcar: TASK [python : Get distribution name from lsb-release] *************************
    sig-flatcar: fatal: [default]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: Unable to negotiate with 127.0.0.1 port 34989: no matching host key type found. Their offer: ssh-rsa", "unreachable": true}`

This was with Flatcar version `3139.2.1`.

With my previous workaround, I still get:

`    sig-flatcar: fatal: [default]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via scp: bash: line 1: /usr/lib/sftp-server: No such file or directory\nscp: Connection closed\r\n", "unreachable": true}`

This was with Flatcar version `3033.2.4`.


I do have `/usr/lib/sftp-server` locally. Trying with Flatcar version

**Comment by kopiczko** [comment]
> I'm testing building and I get this error right now: ` sig-flatcar: TASK [python : Get distribution name from lsb-release] ************************* sig-flatcar: fatal: [default]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: Unable to negotiate with 127.0.0.1 port 34989: no matching host key type found. Their offer: ssh-rsa", "unreachable": true}`
> 
> This was with Flatcar version `3139.2.1` and `3033.2.4`.

Both those versions build fine for me :(

File: images/capi/ansible/roles/node/tasks/main.yml [hunk 4]
```diff
@@ -112,6 +112,6 @@
 
 - name: Copy etcd network tuning script
   copy:
-    src: usr/local/bin/etcd-network-tuning.sh
-    dest: /usr/local/bin/etcd-network-tuning.sh
-    mode: 0755
\ No newline at end of file
+    src: opt/bin/etcd-network-tuning.sh
+    dest: /opt/bin/etcd-network-tuning.sh
```
[the remark this answers is not in the document]
Comment (kopiczko) [reply images/capi/ansible/roles/node/tasks/main.yml:116]: The only reason to do that for all the distros is to reduce complexity. I don't think there is anything wrong in putting scripts in `/opt/bin` in general but if there is please LMK.

File: images/capi/packer/files/bootstrap-flatcar.sh (hunk 1 above)
Comment (kopiczko) [reply images/capi/packer/files/bootstrap-flatcar.sh:5]: You're right this part can be moved to a separate PR. I'll do that later.

There were two issues I had:

- PYTHON_VERSION was not defined (only PYTHON3_VERSION) and `set -o nounset` helps catching that
- `set -o errexit` is the same as `set -e` so there is no change in that regard
- I think `set -o pipefail` just makes sense with `set -o errexit`

I also added:

- `set -x` - to make the debugging easier
- replaced `wget -O -` with `curl -sfL` to reduce the amount of output , because wget is printing all those dots and it's difficult to control that

Comment (kopiczko) [reply images/capi/packer/files/bootstrap-flatcar.sh:5]: BTW I never tested the thing with with the original version of the script successfully. Possibly renaming `PYTHON_VERSION` to `PYTHON3_VERSION` is essential.

File: images/capi/packer/qemu/packer.json (hunk 2 above)
Comment (kopiczko) [reply images/capi/packer/qemu/packer.json:66]: I think yes, but I'll double check that.

File: images/capi/packer/qemu/packer.json (hunk 3 above)
Comment (kopiczko) [reply images/capi/packer/qemu/packer.json:144]: Target VM.

Comment (invidian) [reply images/capi/packer/qemu/packer.json:144]: So I guess it must be copied here, so it can be overwritten by the user? I guess there is no concept of `nil` value which would make it fallback to default in the provisioner?

File: images/capi/ansible/roles/node/tasks/main.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (invidian) [reply images/capi/ansible/roles/node/tasks/main.yml:116]: If we put some other scripts on the image, I'd just follow what they do. If not, I'm fine with it so I'll leave the decision for other maintainers.

**Comment by invidian** [comment]
>Both those versions build fine for me :(

I'll do some more testing to try to figure out what's going on :+1:

File: images/capi/ansible/roles/node/files/etc/udev/rules.d/90-etcd-tuning.rules [hunk 5]
```diff
@@ -12,4 +12,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-ACTION=="add", SUBSYSTEM=="net", SUBSYSTEMS=="pci|xen|vmbus" RUN+="/usr/local/bin/etcd-network-tuning.sh $name"
\ No newline at end of file
+ACTION=="add", SUBSYSTEM=="net", SUBSYSTEMS=="pci|xen|vmbus" RUN+="/opt/bin/etcd-network-tuning.sh $name"
```
[the remark this answers is not in the document]
Comment (invidian) [reply images/capi/ansible/roles/node/files/etc/udev/rules.d/90-etcd-tuning.rules]: It's fixed so can be marked as resolved.

File: images/capi/packer/qemu/packer.json [hunk 6]
```diff
@@ -52,15 +52,18 @@
     },
     {
       "ansible_env_vars": [
-        "ANSIBLE_SSH_ARGS='{{user `existing_ansible_ssh_args`}} -o IdentitiesOnly=yes'"
+        "ANSIBLE_SSH_ARGS='{{user `existing_ansible_ssh_args`}} -o IdentitiesOnly=yes -o HostKeyAlgorithms=+ssh-rsa -o PubkeyAcceptedAlgorithms=+ssh-rsa'"
```
Comment (invidian) [inline images/capi/packer/qemu/packer.json:55]: I figured why it fails to me. We should make it consistent for all providers :) I've been testing with Azure.
```console
$ git grep IdentitiesOnly=yes
packer/ami/packer.json:        "ANSIBLE_SSH_ARGS='{{user `existing_ansible_ssh_args`}} -o IdentitiesOnly=yes'"
packer/azure/packer.json:        "ANSIBLE_SSH_ARGS='{{user `existing_ansible_ssh_args`}} -o IdentitiesOnly=yes'"
packer/azure/packer.json:        "ANSIBLE_SSH_ARGS='{{user `existing_ansible_ssh_args`}} -o IdentitiesOnly=yes'"
packer/digitalocean/packer.json:        "ANSIBLE_SSH_ARGS='{{user `existing_ansible_ssh_args`}} -o IdentitiesOnly=yes'"
packer/gce/packer.json:        "ANSIBLE_SSH_ARGS='{{user `existing_ansible_ssh_args`}} -o IdentitiesOnly=yes'"
packer/oci/packer.json:        "ANSIBLE_SSH_ARGS='{{user `existing_ansible_ssh_args`}} -o IdentitiesOnly=yes'"
packer/ova/packer-node.json:        "ANSIBLE_SSH_ARGS='{{user `existing_ansible_ssh_args`}} -o IdentitiesOnly=yes'"
packer/ova/packer-node.json:        "ANSIBLE_SSH_ARGS='{{user `existing_ansible_ssh_args`}} -o IdentitiesOnly=yes'"
packer/qemu/packer.json:        "ANSIBLE_SSH_ARGS='{{user `existing_ansible_ssh_args`}} -o IdentitiesOnly=yes -o HostKeyAlgorithms=+ssh-rsa -o PubkeyAcceptedAlgorithms=+ssh-rsa'"
packer/raw/packer.json:        "ANSIBLE_SSH_ARGS='{{user `existing_ansible_ssh_args`}} -o IdentitiesOnly=yes'"
```

File: images/capi/ansible/roles/node/files/opt/bin/etcd-network-tuning.sh [hunk 7]
```diff
@@ -24,7 +24,7 @@ on_exit() {
 }
 
 if [ "$#" -ne 1 ]; then
-    echo "Error: Usage: /usr/local/bin/etcd-network-tuning.sh <dev>" | systemd-cat -p emerg -t etcd-tuning
+    echo "Error: Usage: /opt/bin/etcd-network-tuning.sh <dev>" | systemd-cat -p emerg -t etcd-tuning
```
[the remark this answers is not in the document]
Comment (invidian) [reply images/capi/ansible/roles/node/files/opt/bin/etcd-network-tuning.sh]: This is also addressed so can be marked as resolved :)

File: images/capi/packer/qemu/packer.json (hunk 6 above)
Comment (invidian) [reply images/capi/packer/qemu/packer.json:55]: I suggest we deduplicate `-o IdentitiesOnly=yes` to a single place, then add `-o HostKeyAlgorithms=+ssh-rsa -o PubkeyAcceptedAlgorithms=+ssh-rsa` there.

Comment (kopiczko) [reply images/capi/packer/qemu/packer.json:55]: Oh sorry for that 🙈

File: images/capi/packer/qemu/packer.json (hunk 2 above)
Comment (kopiczko) [reply images/capi/packer/qemu/packer.json:66]: When the flag is removed it fails with:

```
    qemu: TASK [Gathering Facts] *********************************************************
    qemu: fatal: [default]: FAILED! => {"msg": "failed to transfer file to /Users/kopiczko/.ansible/tmp/ansible-local-81053co1g_u6h/tmpi5vs4ihs /tmp/.ansible/ansible-tmp-1655402633.0043058-81056-213368382054636/AnsiballZ_setup.py:\n\n"}
```

File: images/capi/packer/qemu/qemu-flatcar.json [hunk 8]
```diff
@@ -1,5 +1,6 @@
 {
   "ansible_extra_vars": "ansible_python_interpreter=/opt/bin/python",
+  "ansible_scp_extra_args": "'-O'",
```
Comment (kopiczko) [inline images/capi/packer/qemu/qemu-flatcar.json:3]: Interestingly I had to set that for ubuntu build as well. @invidian can you check that?

File: images/capi/packer/qemu/packer.json (hunk 3 above)
Comment (kopiczko) [reply images/capi/packer/qemu/packer.json:144]: Empty string worked. I'll replace that.

File: images/capi/packer/qemu/packer.json (hunk 6 above)
Comment (kopiczko) [reply images/capi/packer/qemu/packer.json:55]: I copied that to ami/azure/raw (because only those support Flatcar). Would you like to make a further improvements there?

**Comment by kopiczko** [comment]
CC @johananl

File: images/capi/packer/files/bootstrap-flatcar.sh (hunk 1 above)
Comment (kopiczko) [reply images/capi/packer/files/bootstrap-flatcar.sh:5]: I extracted it here https://github.com/kubernetes-sigs/image-builder/pull/912 after it's merge I can rebase this and this file is gone.

File: images/capi/ansible/roles/node/tasks/main.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (invidian) [reply images/capi/ansible/roles/node/tasks/main.yml:122]: Created https://github.com/kubernetes-sigs/image-builder/pull/914 then :) :broom:

File: images/capi/packer/qemu/packer.json (hunk 6 above)
Comment (invidian) [reply images/capi/packer/qemu/packer.json:55]: I think ssh-rsa key problem is not something Flatcar-specific. I'd expect https://github.com/kubernetes-sigs/image-builder/issues/905 to be a problem everywhere where OpenSSH 9.0+ is used. But maybe the Ansible provisioner for Packer does not setup the proxy in all environments?

I'd rather put all those options in a single place, then add `-o HostKeyAlgorithms=+ssh-rsa -o PubkeyAcceptedAlgorithms=+ssh-rsa` there, as given that this is not a production environment, it shouldn't make much harm and may definitely help with builds reliability.

File: images/capi/packer/qemu/qemu-flatcar.json (hunk 8 above)
Comment (invidian) [reply images/capi/packer/qemu/qemu-flatcar.json:3]: IIUC this is OpenSSH on client side specific, not target specific. At least this was the conclusion from my testing.

Comment (invidian) [reply images/capi/packer/qemu/qemu-flatcar.json:3]: >I wonder if we could update the CI to use OpenSSH 9.0+.

We would have to update CI image to Debian Sid or Bookworm to make this happen, not sure if it's worth it.

File: images/capi/packer/files/bootstrap-flatcar.sh (hunk 1 above)
Comment (johananl) [reply images/capi/packer/files/bootstrap-flatcar.sh:5]: > BTW I never tested the thing with with the original version of the script successfully. Possi

`PYTHON_VERSION` is certainly undefined before your fix, **however** I am able to build despite this fact because when the var reference is interpolated as `""` (empty string), the URL is still valid and points to a PyPy binary 😆

File: images/capi/ansible/roles/node/files/etc/udev/rules.d/90-etcd-tuning.rules (hunk 5 above)
[the remark this answers is not in the document]
Comment (invidian) [reply images/capi/ansible/roles/node/files/etc/udev/rules.d/90-etcd-tuning.rules:15]: @kopiczko can you please create new PR with just this fix? I'd like to unblock other PRs for merging.

**Comment by codenrhoden** [comment]
so what is the exact state of this PR?
It seems to be fixing two problems (the etcd srcipt, and some SSH improvements?) Those should perhaps be two different PRs. This also have 5 commits, some of which are merge commits (and will block the PR).

**Comment by invidian** [comment]
Opened https://github.com/kubernetes-sigs/image-builder/pull/923 with just Flatcar build fix.

**Comment by kopiczko** [comment]
I'd like to have that merged https://github.com/kubernetes-sigs/image-builder/pull/912 and then I could squash it. Or I simply remove the file from this PR

File: images/capi/packer/qemu/packer.json (hunk 6 above)
Comment (invidian) [reply images/capi/packer/qemu/packer.json:55]: Perhaps there is a file with common packer variables across all platforms. Maybe `packer/config/ansible-args.json` ?

File: images/capi/packer/qemu/qemu-flatcar.json (hunk 8 above)
Comment (invidian) [reply images/capi/packer/qemu/qemu-flatcar.json:3]: I guess it should not harm :+1: Having it in one place will also make it easier to verify.

Comment (kopiczko) [reply images/capi/packer/qemu/qemu-flatcar.json]: I put it in `packer/config/ansible-args.json` and added everywhere.

File: images/capi/packer/qemu/packer.json (hunk 6 above)
Comment (kopiczko) [reply images/capi/packer/qemu/packer.json]: I put it there.

**Comment by invidian** [comment]
I'd expect here at least 3 commits though. One for refactoring the parameters to a common place, one with `ssh-rsa` fix referring to appropriate issue, another one for SFTP fix. As merge commits are used in this repository, I think it would be valuable to keep things separate and small, even for the case if some patch needs to be reverted later or for bisecting.

File: images/capi/packer/qemu/packer.json (hunk 2 above)
Comment (invidian) [reply images/capi/packer/qemu/packer.json:94]: Ooh, okay, I see now. This is only needed for Flatcar actually. Make sense.

File: images/capi/packer/config/ansible-args.json [hunk 9]
```diff
@@ -1,3 +1,5 @@
...
+  "ansible_scp_extra_args": "-O",
+  "ansible_ssh_args": "-o IdentitiesOnly=yes -o HostKeyAlgorithms=+ssh-rsa -o PubkeyAcceptedAlgorithms=+ssh-rsa"
```
Comment (invidian) [inline images/capi/packer/config/ansible-args.json:4]: Do we want to use `ansible_common` as a prefix for all variables here?
```suggestion
  "ansible_common_ssh_args": "-o IdentitiesOnly=yes -o HostKeyAlgorithms=+ssh-rsa -o PubkeyAcceptedAlgorithms=+ssh-rsa"
```

File: images/capi/packer/qemu/qemu-flatcar.json [hunk 10]
```diff
@@ -17,6 +17,7 @@
   "os_display_name": "Flatcar Container Linux ({{env `FLATCAR_CHANNEL`}} channel release {{env `FLATCAR_VERSION`}})",
   "python_path": "/opt/bin/builder-env/site-packages",
   "release_version": "{{env `FLATCAR_VERSION`}}",
+  "sftp_command": "/usr/lib64/misc/sftp-server -e",
```
Comment (invidian) [inline images/capi/packer/qemu/qemu-flatcar.json:20]: Hm, I gave it a try and building works fine for me without this. I think we should figure out under which circumstances this is actually needed.

Also adding it does not change anything for me it seems.

**Review by invidian — COMMENTED** [review]
Looks good to me, though I would find it valuable if work could be split into few commits, each addressing specific area, as described above. Refactor, adding workaround for `ssh-rsa` keys, and SFTP issue seem like 3 separate actions to me.

Regarding SFTP command, it seems this part is not required for me, at least for QEMU builds. I'll test later again with SIG builds to verify that.

File: images/capi/packer/qemu/packer.json (hunk 6 above)
Comment (invidian) [reply images/capi/packer/qemu/packer.json]: Looks good, you can mark it as resolved :)

File: images/capi/packer/config/ansible-args.json (hunk 9 above)
Comment (aniruddha2000) [reply images/capi/packer/config/ansible-args.json]:
```json
"ansible_scp_extra_args": "-O",
```
With this I got something like this `ansible-playbook: error: argument --scp-extra-args: expected one argument`
Changing to below works perfectly for GCP
```json
"ansible_scp_extra_args": "\"-O\"",
```

File: images/capi/packer/qemu/qemu-flatcar.json (hunk 10 above)
Comment (kopiczko) [reply images/capi/packer/qemu/qemu-flatcar.json:20]: I tried it and and works for me without it too... 🤯

Comment (kopiczko) [reply images/capi/packer/qemu/qemu-flatcar.json]: Removed

File: images/capi/packer/config/ansible-args.json (hunk 9 above)
Comment (invidian) [reply images/capi/packer/config/ansible-args.json]: Yeah, same happens to me:
```console
$ ansible --version
ansible [core 2.11.5]
  config file = /home/invidian/data/workspaces/clusterapi-flatcar/image-builder/images/capi/ansible.cfg
  configured module search path = ['/home/invidian/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/invidian/.local/lib/python3.10/site-packages/ansible
  ansible collection location = /home/invidian/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/invidian/.local/bin/ansible
  python version = 3.10.5 (main, Jun  6 2022, 18:49:26) [GCC 12.1.0]
  jinja version = 3.1.2
  libyaml = True
```

Comment (aniruddha2000) [reply images/capi/packer/config/ansible-args.json]: @kopiczko 
```bash
ansible [core 2.12.7]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/aniruddha/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /home/aniruddha/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.10.4 (main, Apr  2 2022, 09:04:19) [GCC 11.2.0]
  jinja version = 3.0.3
  libyaml = True
```

File: images/capi/packer/qemu/qemu-flatcar.json (hunk 10 above)
Comment (kopiczko) [reply images/capi/packer/qemu/qemu-flatcar.json]: I think it's because of the `--scp-extra-args`. When it's provided then ansible switches from sftp to scp... I'm not sure if this is what we want. That's why we were confused why both are required at the beginning.

**Comment by invidian** [comment]
The commit message is not aligned with it's content anymore.

File: images/capi/packer/qemu/qemu-flatcar.json (hunk 10 above)
Comment (kopiczko) [reply images/capi/packer/qemu/qemu-flatcar.json]: It looks like the newest `scp` versions [use SFTP under the hood](https://www.redhat.com/en/blog/openssh-scp-deprecation-rhel-9-what-you-need-know):

```
        -O Use the legacy SCP protocol for file transfers instead of the
           SFTP protocol. Forcing the use of the SCP protocol may be
           necessary for servers that do not implement SFTP, for
           backwards-compatibility for particular filename wildcard
           patterns and for expanding paths with a ‘~’ prefix for older
           SFTP servers.
```

So it isn't that `--scp-extra-args` alone switches the file transfer protocol, but it's the actual `-O`.

**Comment by kopiczko** [comment]
@codenrhoden @invidian from my POV this is ready to be merged. I tested it, split into cherry-pickable commits and rebased.

@aniruddha2000 I added quotation around `-O`. It was also failing for me. I didn't notice in the first place because of my dev branch quirk.

File: images/capi/packer/config/ansible-args.json (hunk 9 above)
Comment (kopiczko) [reply images/capi/packer/config/ansible-args.json]: Added quotations. It also fails for me. I didn't notice in the first place because of my overwrites in my dev branch.

File: images/capi/packer/config/ansible-args.json [hunk 11]
```diff
@@ -1,3 +1,5 @@
...
+  "ansible_common_ssh_args": "-o IdentitiesOnly=yes -o HostKeyAlgorithms=+ssh-rsa -o PubkeyAcceptedAlgorithms=+ssh-rsa",
+  "ansible_scp_extra_args": "\"-O\""
```
Comment (invidian) [inline images/capi/packer/config/ansible-args.json:4]: Single quotes would also work, not sure which variant is actually more readable though.
```suggestion
  "ansible_scp_extra_args": "'-O'"
```

**Review by invidian — APPROVED** [review]
I haven't tested it now, but did it before and it was OK, so LGTM, thanks for getting back to it @kopiczko!

File: images/capi/packer/config/ansible-args.json (hunk 11 above)
Comment (kopiczko) [reply images/capi/packer/config/ansible-args.json]: Changed

**Comment by fabriziopandini** [comment]
I don't have enough context about imageBuilder to validate a change impacting so many providers 😓 
It would be great to have folks with specific expertise involved instead
/unassign

**Review by CecileRobertMichon — COMMENTED** [review]
/ok-to-test
/cc @kkeshavamurthy @jsturtevant

**Comment by jsturtevant** [comment]
looks like the failures are due to some of the changes with ssh, both 1804 and 2004 are failing with:

```
[0;32m    vhd-ubuntu-2004: TASK [include_role : python] ***************************************************[0m
[0;32m    vhd-ubuntu-2004: [WARNING]: raw module does not support the environment keyword[0m
[0;32m    vhd-ubuntu-2004:[0m
[0;32m    vhd-ubuntu-2004: TASK [python : Get distribution name from lsb-release] *************************[0m
[0;32m    vhd-ubuntu-2004: fatal: [default]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: command-line: line 0: Bad configuration option: pubkeyacceptedalgorithms", "unreachable": true}[0m
[0;32m    vhd-ubuntu-2004:[0m
[0;32m    vhd-ubuntu-2004: PLAY RECAP *********************************************************************[0m
[0;32m    vhd-ubuntu-2004: default                    : ok=0    changed=0    unreachable=1    failed=0    skipped=0    
```

**Comment by kkeshavamurthy** [comment]
@kopiczko do you have any updates on this?

**Comment by kopiczko** [comment]
CI is failing with:

```
[0;32m    vsphere-clone: fatal: [default]: FAILED! => {"msg": "failed to transfer file to /root/.ansible/tmp/ansible-local-1386oe5wt7mv/tmpx_roh_r0 /tmp/.ansible/ansible-tmp-1661355877.2552273-1392-127598361667597/AnsiballZ_setup.py:\n\nunknown option -- O\r\nusage: scp [-346BCpqrv] [-c cipher] [-F ssh_config] [-i identity_file]\n           [-l limit] [-o ssh_option] [-P port] [-S program] source ... target\n"}[0m
```

So this is the host issue. It looks like the CI VMs there are running some older OpenSSH version. I'm not sure how to proceed from here. Should I simply close it? Should there at least be some docs on how to build with newer OpenSSH?

**Comment by invidian** [comment]
>So this is the host issue. It looks like the CI VMs there are running some older OpenSSH version. I'm not sure how to proceed from here. Should I simply close it? Should there at least be some docs on how to build with newer OpenSSH?

As I mentioned in https://github.com/kubernetes-sigs/image-builder/pull/907#discussion_r901413880, CI runs in Docker container running Debian. Or at least have been at this point in time.

**Comment by aniruddha2000** [comment]
Are there any updates on this? We are waiting on this to merge so that we can finalize our [PR](https://github.com/kubernetes-sigs/image-builder/pull/934) 😅

**Comment by kopiczko** [comment]
> Are there any updates on this? We are waiting on this to merge so that we can finalize our [PR](https://github.com/kubernetes-sigs/image-builder/pull/934) 😅

@aniruddha2000 unfortunately I don't think it's going anywhere without CI OS being upgraded. Can you describe why it is blocking your PR?

**Comment by kkeshavamurthy** [comment]
> CI is failing with:
> 
> ```
> �[0;32m    vsphere-clone: fatal: [default]: FAILED! => {"msg": "failed to transfer file to /root/.ansible/tmp/ansible-local-1386oe5wt7mv/tmpx_roh_r0 /tmp/.ansible/ansible-tmp-1661355877.2552273-1392-127598361667597/AnsiballZ_setup.py:\n\nunknown option -- O\r\nusage: scp [-346BCpqrv] [-c cipher] [-F ssh_config] [-i identity_file]\n           [-l limit] [-o ssh_option] [-P port] [-S program] source ... target\n"}�[0m
> ```
> 
> So this is the host issue. It looks like the CI VMs there are running some older OpenSSH version. I'm not sure how to proceed from here. Should I simply close it? Should there at least be some docs on how to build with newer OpenSSH?

@kopiczko the jobs use the [latest](https://github.com/kubernetes/test-infra/blame/c3af09ab20d6faebd27ac6a5000f0d15dde810a8/config/jobs/kubernetes-sigs/image-builder/image-builder-ova-presubmits.yaml#L16) container images to run the test run. 
If we do not need latest openSSH package on it, maybe we can PR that somewhere?

**Comment by aniruddha2000** [comment]
@kopiczko  Our `ansible_env_vars` were failing but with your changes in this PR, we successfully created our OS image.
So if your ansible changes are merged in this PR we will merge our changes too.

**Comment by kralicky** [comment]
These changes fixed the following qemu image build errors for me. :+1: 
`fatal: [default]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: Unable to negotiate with 127.0.0.1 port 46489: no matching host key type found. Their offer: ssh-rsa", "unreachable": true}`
and
`fatal: [default]: FAILED! => {"msg": "failed to transfer file to /home/joe/.ansible/tmp/ansible-local-1545693264eeguh/tmpmq7bxkpk /tmp/.ansible/ansible-tmp-1666398201.818839-1545697-148037159050528/AnsiballZ_setup.py:\n\n"}`

**Comment by kopiczko** [comment]
> > CI is failing with:
> > ```
> > �[0;32m    vsphere-clone: fatal: [default]: FAILED! => {"msg": "failed to transfer file to /root/.ansible/tmp/ansible-local-1386oe5wt7mv/tmpx_roh_r0 /tmp/.ansible/ansible-tmp-1661355877.2552273-1392-127598361667597/AnsiballZ_setup.py:\n\nunknown option -- O\r\nusage: scp [-346BCpqrv] [-c cipher] [-F ssh_config] [-i identity_file]\n           [-l limit] [-o ssh_option] [-P port] [-S program] source ... target\n"}�[0m
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So this is the host issue. It looks like the CI VMs there are running some older OpenSSH version. I'm not sure how to proceed from here. Should I simply close it? Should there at least be some docs on how to build with newer OpenSSH?
> 
> @kopiczko the jobs use the [latest](https://github.com/kubernetes/test-infra/blame/c3af09ab20d6faebd27ac6a5000f0d15dde810a8/config/jobs/kubernetes-sigs/image-builder/image-builder-ova-presubmits.yaml#L16) container images to run the tests. If we do not have the latest openSSH package on it, maybe we can PR that somewhere?

@kkeshavamurthy do you know where that somewhere is?

**Comment by kkeshavamurthy** [comment]
@kopiczko Sorry for the very delayed reply. 
I would guess [test-infra](https://github.com/kubernetes/test-infra) repo is where the container images are built. Maybe reach out their slack channel to get more info?

**Comment by johananl** [comment]
Likely location of the source files used to build the kubekins-e2e image: https://github.com/kubernetes/test-infra/tree/master/images/kubekins-e2e

Likely relevant Slack channel: [#sig-k8s-infra](https://kubernetes.slack.com/archives/CCK68P2Q2)

**Comment by johananl** [comment]
So the latest kubekins-e2e image (which is the image used by the CI for image builds) is based on Debian Buster which has openssh-client `7.9p1`. That's why the CI doesn't recognize the `-O` scp flag.

**Comment by johananl** [comment]
@kopiczko I think we need to rename this PR because it's not Flatcar-related. True, right now the problem influences Flatcar only but that's because Flatcar already has **openssh >=9.0**. The problem will gradually affect the rest of the distros so we better fix it in advance.

**Comment by johananl** [comment]
So I guess the plan is:

- Find a way to update openssh in the CI images so that we can use the `-O` SCP workaround.
- Leave the Packer var override which **disables** the workaround for any distro which still uses an older openssh.
- Gradually remove the override as distros start to use openssh >= 9.0 so that eventually all distros use the `-O` option.

**Comment by invidian** [comment]
So if CI container does not support `scp -O` flag, does `scp` works just fine for building Flatcar then?

I think it would be fine if we just document that if you run OpenSSH 9.0+ locally, then you need an extra variable with `ansible_scp_extra_args: "-O",` or something, until it becomes the default when CI is updated.

**Comment by invidian** [comment]
Wait, but CI is right now passing for Flatcar. So building only fails if your local SSH version is too new I guess? In such case, we could add some logic to conditionally add `-O` flag to SCP I think?

@kopiczko mind if I open another PR with those changes? I'd like to get this merged.

**Comment by kopiczko** [comment]
@invidian please do. Our priority shifted currently and I don't have much time and focus to work on that now. LMK if I should close that one.

**Comment by invidian** [comment]
Thanks for the response. I think I figured an acceptable way forward. I'll open a PR tomorrow and I think we can then close this one.

**Comment by invidian** [comment]
Created https://github.com/kubernetes-sigs/image-builder/pull/1035 then.

**Comment by kopiczko** [comment]
Thanks @invidian, closing this one.
